borg/archiver.py

```diff
@@ -353,6 +355,155 @@ def do_extract(self, args):
                 self.print_warning("Include pattern '%s' never matched.", pattern)
         return self.exit_code
 
+    def do_diff(self, args):
+        """Diff archive contents"""
+        def dump_json(dictionary):
+            if dump_json.wants_comma:
+                sys.stdout.write(",")
+            dump_json.wants_comma = True
+            json.dump(dictionary, sys.stdout)
+        dump_json.wants_comma = False
+
+        def format_bytes(count):
+            if count is None:
+                return "<deleted>"
+            return format_file_size(count)
+
+        def compare_chunks(lhs, rhs, lhs_archive, rhs_archive):
+            lhs_hash = sha256()
+            rhs_hash = sha256()
+            for chunk in lhs_archive.pipeline.fetch_many(lhs):
+                lhs_hash.update(chunk)
+            for chunk in rhs_archive.pipeline.fetch_many(rhs):
+                rhs_hash.update(chunk)
+            return lhs_hash.digest() == rhs_hash.digest()
```

Comment:
So this is simple and all, but inefficient for large files, especially if they changed early in the file. I tried an alternate implementation in Python, basically comparing prefixes of chunks with a moving `index` (i.e. larger chunk starts with smaller chunk, get next chunk, do index starting at offset), but it was overall slower and way more complicated. I also don't think I caught all of the corner cases.

So I opted for this simple implementation where it is easy to see that it does the right thing. I also doubt many people actually change chunker params at all, so this fallback path is rather unimportant I think.